### PR TITLE
Use ContextResourceString to use context property

### PR DESCRIPTION
**Pull request by gouniLee**
In QML apps, If the same source/key is duplicated in different files a project, loctool should write them differently. So I update to code to use `ContextResourceString` instead `ResourceString`. It uses context info when making a key/hash key. then I use file info as context.

I need to do additional work to match with xliff unit data.  Because xliff data doesn't have context info so don't match the hashkey from ContextResourceString

### CODE REVIEW

**Comment by gouniLee** [comment]
It has `ilib-loctool-webos-ts-resource` module dependency. but it has not been published to npm yet. That's why travis is failed. It passed all on my local machine.

File: QMLFile.js [hunk 1]
```diff
@@ -21,6 +21,7 @@ var fs = require("fs");
 var path = require("path");
 var log4js = require("log4js");
 var logger = log4js.getLogger("loctool.plugin.QMLFile");
+var ContextResourceString = require("loctool/lib/ContextResourceString.js");
```
Comment (ehoogerbeets) [inline QMLFile.js]: You cannot depend on the loctool in the plugin for the loctool because that creates a circular dependency. You have to create resources with this.API.newResource()

File: QMLFileType.js [hunk 2]
```diff
@@ -21,6 +21,7 @@ var fs = require("fs");
 var log4js = require("log4js");
 var QMLFile = require("./QMLFile.js");
 var TSResourceFileType = require("ilib-loctool-webos-ts-resource");
+var ContextResourceString = require("loctool/lib/ContextResourceString");
```
Comment (ehoogerbeets) [inline QMLFileType.js]: Same here

File: QMLFile.js [hunk 3]
```diff
@@ -127,8 +134,7 @@ QMLFile.prototype.parse = function(data) {
             comment = commentArr.join(" ");
 
             match = QMLFile.unescapeString(match);
-
-            var r = this.API.newResource({
```
Comment (ehoogerbeets) [inline QMLFile.js]: In order to change this factory method `this.API.newResource` for the string resType, you have to implement the `getResourceTypes` method in the `QmlFileType` plugin to tell it that "string" means `ContextResourceString`. It looks like you already did this, so you just need to call `this.API.newResource` as normal.

File: QMLFileType.js [hunk 4]
```diff
@@ -92,9 +93,11 @@ QMLFileType.prototype.write = function(translations, locales) {
         // for each extracted string, write out the translations of it
         translationLocales.forEach(function(locale) {
             logger.trace("Localizing QML strings to " + locale);
+            var manipulateKey = res.cleanHashKeyForTranslation(locale).replace(res.getContext(),"");
```
Comment (ehoogerbeets) [inline QMLFileType.js]: Why did you delete the context from the hash for the context string? I think that is the whole point for using a context string instead of just a regular string...

File: QMLFileType.js [hunk 5]
```diff
@@ -118,11 +121,14 @@ QMLFileType.prototype.write = function(translations, locales) {
                         r = r.clone();
                         r.reskey = res.reskey;
                     }
+                    var storeResoure = r.clone();
```
Comment (ehoogerbeets) [inline QMLFileType.js]: should be `var storeResource`

File: test/testQMLFile.js [hunk 6]
```diff
@@ -21,7 +21,7 @@ if (!QMLFile) {
     var QMLFile = require("../QMLFile.js");
     var QMLFileType = require("../QMLFileType.js");
     var CustomProject =  require("loctool/lib/CustomProject.js");
-    var ResourceString =  require("loctool/lib/ResourceString.js");
+    var ContextResourceString =  require("loctool/lib/ContextResourceString.js");
```
Comment (ehoogerbeets) [inline test/testQMLFile.js]: The tests can depend on the loctool because the loctool does not depend on the tests so there is no circular dependency.

File: QMLFileType.js (hunk 4 above)
Comment (gouniLee) [reply QMLFileType.js]: When the loctool parse the extracted string, the hashkey is something like below:
 ` crs_app1_Volume_ko-KR_hello_x-qml_, `
(Volume is context value, hello is source string)

But, When the parsing xliff files, It stores a value with key, but at this time xliff file doesn't have context value.

So matched target string's key is
'crs_app1__ko-KR_hello'

So In order to find proper translated value from xliff, the key shouldn't contain context value. If not, I can't get the correct translated string.

File: QMLFile.js (hunk 3 above)
Comment (gouniLee) [reply QMLFile.js]: Yeah..  It should use `this.API.newResource` I'll update it.
If I modified it, it works ok with running real App, Because ContexResource setting work doing under Project's init() stage. https://github.com/iLib-js/loctool/blob/development/lib/Project.js#L143  
But the unit test cases, It directly linked ResourceFactory, https://github.com/iLib-js/loctool/blob/development/lib/ResourceFactory.js#L62
I'm stuck how to properly set ContextResourceString instead ResourceString.

File: QMLFileType.js (hunk 4 above)
Comment (ehoogerbeets) [reply QMLFileType.js]: Do you mean that the current LG xliffs don't have this context? If not, maybe you make the code check for the string with the context, and if it is not there, check for it without the context.

File: QMLFile.js (hunk 3 above)
Comment (ehoogerbeets) [reply QMLFile.js]: You already did that correctly in `QMLFileType.prototype.getResourceTypes()`

File: QMLFileType.js (hunk 4 above)
Comment (gouniLee) [reply QMLFileType.js]: Yes, LG xliffs don't have context name. (Ih has only source, target, and comment). I updated code to check two times to get a translation. One with a full key, another is key without context.

File: QMLFile.js (hunk 3 above)
Comment (gouniLee) [reply QMLFile.js]: No, It needs more work to make all unit test pass itself. Please see the latest Travis result)
(Even though module works ok when  running with real an app)

**Comment by gouniLee** [comment]
@ehoogerbeets, 
In order to initialize a custom project, It needs to load plugin `ilib-loctool-webos-qml` itself.
So I've additionally add that module's directory under test/ and For loading that Plugin correctly, loctool's Customproject.js  file need to add that path something like:  https://github.com/iLib-js/loctool/commit/967f1efa2fb3de728df783083af1fc5db0aab2aa
I've checked it passed all the test cases with these changes. Do you think it's an appropriate modification?

**Comment by gouniLee** [comment]
If I give ["../."] as value for plugins when creating CustomProject, the test is finished with giving the following message..
```
15:35:10 INFO loctool.loctool: Command: localize
15:35:10 INFO loctool.loctool: Searching root: .

15:35:10 INFO loctool.loctool: ilib-loctool-webos-qml
15:35:10 INFO loctool.loctool: testfiles
15:35:10 INFO loctool.loctool: Done

FAILURES: Undone tests (or their setups/teardowns): 
- testQMLFileConstructor
```
I expected to load default `package.json` then load qml module. But
When I debug,  the process is done when running line https://github.com/iLib-js/loctool/blob/development/lib/CustomProject.js#L76
Can you tell me know the working examples to check as a reference?

**Comment by gouniLee** [comment]
@ehoogerbeets oh! Thank you for the fix. :sweat_smile: would it be ok to merge?
